### PR TITLE
Add multithread example

### PR DESCRIPTION
**Pull request by Acuadros95**

### CODE REVIEW

File: rclc/multithread_publisher_subscriber/main.c [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+	RCCHECK(rclc_node_init_default(&node, "multithread_node", "", &support));
+
+	// Create publishers
+	for (size_t i = 0; i < PUBLISHER_NUMBER; i++)
+	{
+		RCCHECK(rclc_publisher_init_default(
+			&publisher[i],
+			&node,
+			ROSIDL_GET_MSG_TYPE_SUPPORT(std_msgs, msg, Header),
+			"/microROS/multithread"));
```
Comment (pablogs9) [inline rclc/multithread_publisher_subscriber/main.c:90]:
```suggestion
			"multithread_publisher"));
```

File: rclc/multithread_publisher_subscriber/main.c [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+	}
+
+	// Create subscriber
+	RCCHECK(rclc_subscription_init_default(
+		&subscriber,
+		&node,
+		ROSIDL_GET_MSG_TYPE_SUPPORT(std_msgs, msg, Header),
+		"/microROS/multithread"));
```
Comment (pablogs9) [inline rclc/multithread_publisher_subscriber/main.c:98]:
```suggestion
		"multithread_subscriber"));
```

File: rclc/multithread_publisher_subscriber/main.c [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+	
+	// Allocate the subscriber message
+	char recv_msg_buffer[STRING_BUFFER_LEN];
```
Comment (pablogs9) [inline rclc/multithread_publisher_subscriber/main.c:101]: Init the msg with the new util: https://github.com/micro-ROS/micro-ROS-demos/blob/6cba26276e283b598a5e82c93d365b3709895d0c/rclc/static_type_handling/main.c#L71 
And give feedback about it.

File: rclc/CMakeLists.txt [hunk 4]
```diff
@@ -57,6 +57,10 @@ export_executable(epoch_synchronization)
 export_executable(parameter_server)
 # export_executable(configuration_example custom_transports)
 
+if(MULTITHREAD_EXAMPLE)
+    export_executable(multithread_publisher_subscriber)   
+endif()
```
Comment (Acuadros95) [inline rclc/CMakeLists.txt:62]:
```suggestion
export_executable(multithread_publisher_subscriber)
```

File: rclc/multithread_publisher_subscriber/main.c [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+	{
+		struct arg_struct *args = malloc(sizeof(struct arg_struct));
+		args->publisher = &publisher[i];
+		args->index = i;
+
+		pthread_create(&pub_thr[i], NULL, publish, args);
+	}
+
+	// Set executor timeout to 100 ms to reduce thread locking time
+	executor.timeout_ns = 100000000;
```
Comment (pablogs9) [inline rclc/multithread_publisher_subscriber/main.c:121]: You have an API for setting this: https://github.com/ros2/rclc/blob/aed2a7d652073b77f87e35b4dbbbf43405c2b573/rclc/include/rclc/executor.h#L143

File: rclc/multithread_publisher_subscriber/main.c [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+// Publish thread
+void* publish(
+        void* args)
+{
+	struct arg_struct *arguments = args;
+	rcl_publisher_t *publisher = arguments->publisher;
+	uint8_t id = arguments->index;
+
+	// Create and allocate the publisher message
+	std_msgs__msg__Header send_msg;
```
Comment (pablogs9) [inline rclc/multithread_publisher_subscriber/main.c:41]: Init the msg with the new util: https://github.com/micro-ROS/micro-ROS-demos/blob/6cba26276e283b598a5e82c93d365b3709895d0c/rclc/static_type_handling/main.c#L71 
And give feedback about it.

File: rclc/multithread_publisher_subscriber/main.c [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+		RCSOFTCHECK(rcl_publish(publisher, &send_msg, NULL));
+		printf("Thread %d sent: %d\n", id, send_msg.stamp.nanosec);	
+        usleep(period_us);
+    }
+}
+
+void subscription_callback(const void * msgin)
+{
+	const std_msgs__msg__Header * msg = (const std_msgs__msg__Header *)msgin;
+	printf("Received %d from %s\n", msg->stamp.nanosec, msg->frame_id.data);
```
Comment (pablogs9) [inline rclc/multithread_publisher_subscriber/main.c:69]: Print also `msg->stamp.sec`

**Review by pablogs9 — CHANGES_REQUESTED** [review]

File: rclc/multithread_publisher_subscriber/main.c [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+	RCCHECK(rclc_node_init_default(&node, "multithread_node", "", &support));
+
+	// Create publishers
+	for (size_t i = 0; i < PUBLISHER_NUMBER; i++)
+	{
+		RCCHECK(rclc_publisher_init_default(
+			&publisher[i],
+			&node,
+			ROSIDL_GET_MSG_TYPE_SUPPORT(std_msgs, msg, Header),
+			"multithread_publisher"));
```
Comment (pablogs9) [inline rclc/multithread_publisher_subscriber/main.c:90]: I'm not sure if you can create multiple publishers with the same name... Maybe we should call them: `multithread_publisher_1`, `multithread_publisher_2`,...

File: rclc/multithread_publisher_subscriber/main.c (hunk 7 above)
Comment (Acuadros95) [reply rclc/multithread_publisher_subscriber/main.c]: https://github.com/micro-ROS/micro-ROS-demos/blob/9f6efc8f324973cc1538e129697b8ec3774a78ae/rclc/multithread_publisher_subscriber/main.c#L73

https://github.com/micro-ROS/micro-ROS-demos/blob/9f6efc8f324973cc1538e129697b8ec3774a78ae/rclc/multithread_publisher_subscriber/main.c#L81

File: rclc/multithread_publisher_subscriber/main.c (hunk 5 above)
Comment (Acuadros95) [reply rclc/multithread_publisher_subscriber/main.c]: https://github.com/micro-ROS/micro-ROS-demos/blob/9f6efc8f324973cc1538e129697b8ec3774a78ae/rclc/multithread_publisher_subscriber/main.c#L141-L143

**Review by pablogs9 — APPROVED** [review]

File: rclc/multithread_publisher_subscriber/main.c (hunk 3 above)
Comment (Acuadros95) [reply rclc/multithread_publisher_subscriber/main.c]: https://github.com/micro-ROS/micro-ROS-demos/blob/9f6efc8f324973cc1538e129697b8ec3774a78ae/rclc/multithread_publisher_subscriber/main.c#L112-L118

File: rclc/multithread_publisher_subscriber/main.c (hunk 6 above)
Comment (Acuadros95) [reply rclc/multithread_publisher_subscriber/main.c:44]: Applied on: 9f6efc8f324973cc1538e129697b8ec3774a78ae

Worked as expected. It really simplifies the configuration and allocation of message memory.

**Pull request merged**